src/function/vector_cast_functions.cpp

```diff
@@ -600,28 +612,44 @@ static std::unique_ptr<ScalarFunction> bindCastToNumericFunction(const std::stri
         func);
 }
 
-static std::unique_ptr<ScalarFunction> bindCastBetweenNested(const std::string& functionName,
+static std::unique_ptr<ScalarFunction> bindCastToUnionFunction(const std::string& functionName,
     const LogicalType& sourceType, const LogicalType& targetType) {
-    switch (sourceType.getLogicalTypeID()) {
-    case LogicalTypeID::LIST:
-    case LogicalTypeID::MAP:
-    case LogicalTypeID::STRUCT:
-    case LogicalTypeID::ANY:
-    case LogicalTypeID::ARRAY: {
-        // todo: compile time checking of nested types
-        if (CastArrayHelper::checkCompatibleNestedTypes(sourceType.getLogicalTypeID(),
-                targetType.getLogicalTypeID())) {
+    // source type is not nested, targetType is a union
+    uint64_t numFields = UnionType::getNumFields(targetType);
+    for (uint64_t i = 0; i < numFields; ++i) {
+        const LogicalType& fieldType = UnionType::getFieldType(targetType, i);
```

Comment:
Our convension is to use auto as much as possible.
So you can change it to `const auto& fieldType = xxx`